corehq/apps/domain/tests/test_domain_name_generation.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import print_function, unicode_literals
+
+from django.test import SimpleTestCase, TestCase
+
+from corehq.apps.domain.models import Domain
+from corehq.apps.domain.exceptions import NameUnavailableException
+
+
+class DomainNameAvailabilityTest(SimpleTestCase):
+    def test_conflict(self):
+        self.assertEquals(Domain._get_next_available_name("fandango", []), "fandango-1")
+
+    def test_availability(self):
+        name = "abc"
+        similar = []
+
+        self.assertEquals(Domain._get_next_available_name(name, similar), "abc-1")
+
+        similar.append("abc-1")
+        self.assertEquals(Domain._get_next_available_name(name, similar), "abc-2")
+
+        # Don't bother filling in gaps
+        similar.append("abc-9")
+        self.assertEquals(Domain._get_next_available_name(name, similar), "abc-10")
+
+
+class DomainNameGenerationTest(TestCase):
+    def setUp(self):
+        self.domains = []
+
+    def add_domain(self, name):
+        domain = Domain(name=name)
+        domain.save()
+        self.domains.append(domain)
+
+    def tearDown(self):
+        for domain in self.domains:
+            domain.delete()
+
+    def test_generation(self):
+        self.assertEquals(Domain.generate_name("I have  spaces"), "i-have-spaces")
```

Comment:
(not blocking) I think you could also pull out some of the space/lowercase logic into its own function and then make more of these tests not depend on the DB.